auraed/src/cri/runtime_service.rs

```diff
@@ -115,72 +125,111 @@ impl runtime_service_server::RuntimeService for RuntimeService {
-        let syscall = create_syscall();
-        let sandbox_builder = ContainerBuilder::new(
-            AURAE_SELF_IDENTIFIER.to_string(),
-            syscall.as_ref(),
-        );
-
-        // Spawn auraed here
-        // TODO Check if sandbox already exists?
-        let _spawned = spawn_auraed_oci_to(
-            Path::new(AURAE_BUNDLE_PATH)
-                .join(AURAE_SELF_IDENTIFIER)
-                .to_str()
-                .expect("recursive path"),
-            oci_builder.build().expect("building pod oci spec"),
-        );
-
-        let sandbox_id = metadata.name;
-        let mut sandbox = sandbox_builder
-            .with_root_path(Path::new(AURAE_PODS_PATH).join(sandbox_id.clone()))
-            .expect("Setting pods directory")
-            .as_init(Path::new(AURAE_BUNDLE_PATH).join(AURAE_SELF_IDENTIFIER))
-            .with_systemd(false)
-            .build()
-            .expect("failed building pod sandbox: ensure valid OCI spec and proper container starting point");
-
-        sandbox.start().expect("starting pod sandbox");
+        let mut sandbox: Option<Container>;
+        {
+            // Initialize a new container builder with the AURAE_SELF_IDENTIFIER name as the "init" container running a recursive Auraed
+            let syscall = create_syscall();
+            let sandbox_builder = ContainerBuilder::new(
+                AURAE_SELF_IDENTIFIER.to_string(),
+                syscall.as_ref(),
+            );
+
+            // Spawn auraed here
+            // TODO Check if sandbox already exists and if it does, clean up to this point?
+            let _spawned = spawn_auraed_oci_to(
+                Path::new(AURAE_BUNDLE_PATH)
+                    .join(AURAE_SELF_IDENTIFIER)
+                    .to_str()
+                    .expect("recursive path"),
+                oci_builder.build().expect("building pod oci spec"),
+            );
+
+            sandbox = Some(sandbox_builder
+                .with_root_path(Path::new(AURAE_PODS_PATH).join(sandbox_id.clone()))
+                .expect("Setting pods directory")
+                .as_init(Path::new(AURAE_BUNDLE_PATH).join(AURAE_SELF_IDENTIFIER))
+                .with_systemd(false)
+                .build()
+                .expect("failed building pod sandbox: ensure valid OCI spec and proper container starting point"));
+
+            sandbox
+                .as_mut()
+                .expect("a sandbox")
+                .start()
+                .expect("starting pod sandbox");
+        }
 
-        // TODO: Cache sandbox
+        // NOTE: we're creating a libcontainer::container and calling it `sandbox`,
+        // so that's what goes in the cache.  but i think the rest of this service
+        // differentiates between the outer "pod" and the inner "container"s so we
+        // might want to do the same.
+        let mut sandboxes = self.sandboxes.lock().await;
```

Comment:
Since the lock is taken after the sandbox is created, do we have a race condition. The cache will correctly error on the second one, but then we need to clean up.

Reply:
ack.. moved the lock up.